Validation/HGCalValidation/plugins/HeterogeneousHGCalRecHitsValidator.cc

```diff
@@ -0,0 +1,109 @@
+#include "Validation/HGCalValidation/plugins/HeterogeneousHGCalRecHitsValidator.h"
+
+#include "DetectorDescription/OfflineDBLoader/interface/GeometryInfoDump.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/HGCalGeometry/interface/HGCalGeometry.h"
+#include "SimG4CMS/Calo/interface/CaloHitID.h"
+
+#include "DetectorDescription/Core/interface/DDFilter.h"
+#include "DetectorDescription/Core/interface/DDFilteredView.h"
+#include "DetectorDescription/Core/interface/DDSolid.h"
+
+#include "DataFormats/GeometryVector/interface/Basic3DVector.h"
+
+#include "CLHEP/Units/GlobalSystemOfUnits.h"
+
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
+
+HeterogeneousHGCalRecHitsValidator::HeterogeneousHGCalRecHitsValidator(const edm::ParameterSet &ps)
+    : tokens_({{{{consumes<HGCRecHitCollection>(ps.getParameter<edm::InputTag>("cpuRecHitsEEToken")),
+                  consumes<HGCRecHitCollection>(ps.getParameter<edm::InputTag>("gpuRecHitsEEToken"))}},
+                {{consumes<HGCRecHitCollection>(ps.getParameter<edm::InputTag>("cpuRecHitsHSiToken")),
+                  consumes<HGCRecHitCollection>(ps.getParameter<edm::InputTag>("gpuRecHitsHSiToken"))}},
+                {{consumes<HGCRecHitCollection>(ps.getParameter<edm::InputTag>("cpuRecHitsHSciToken")),
+                  consumes<HGCRecHitCollection>(ps.getParameter<edm::InputTag>("gpuRecHitsHSciToken"))}}}}),
+      treenames_({{"CEE", "CHSi", "CHSci"}}) {
+  edm::Service<TFileService> fs;
+  for (unsigned int i = 0; i < nsubdetectors; ++i) {
+    trees_[i] = fs->make<TTree>(treenames_[i].c_str(), treenames_[i].c_str());
+    trees_[i]->Branch("cpu", "validHitCollection", &cpuValidRecHits[i]);
+    trees_[i]->Branch("gpu", "validHitCollection", &gpuValidRecHits[i]);
+    trees_[i]->Branch("diffs", "validHitCollection", &diffsValidRecHits[i]);
+  }
+}
+
+HeterogeneousHGCalRecHitsValidator::~HeterogeneousHGCalRecHitsValidator() {}
+
+void HeterogeneousHGCalRecHitsValidator::endJob() {}
+
+void HeterogeneousHGCalRecHitsValidator::set_geometry_(const edm::EventSetup &setup, const unsigned int &detidx) {
+  edm::ESHandle<HGCalGeometry> handle;
+  setup.get<IdealGeometryRecord>().get(handles_str_[detidx], handle);
+}
+
+void HeterogeneousHGCalRecHitsValidator::analyze(const edm::Event &event, const edm::EventSetup &setup) {
+  edm::ESHandle<CaloGeometry> baseGeom;
+  setup.get<CaloGeometryRecord>().get(baseGeom);
+  recHitTools_.setGeometry(*baseGeom);
+
+  //future subdetector loop
+  for (size_t idet = 0; idet < nsubdetectors; ++idet) {
+    set_geometry_(setup, idet);
+
+    //get hits produced with the CPU
+    event.getByToken(tokens_[idet][0], handles_[idet][0]);
+    const auto &cpuhits = *handles_[idet][0];
```

Comment:
If you don't need the Handle for anything else than getting the product, this could be just
```suggestion
    const auto &cpuhits = event.get(tokens_[idet][0]);
```